source/memory_pool.c

```diff
@@ -0,0 +1,116 @@
+/*
+* Copyright 2010-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+*
+* Licensed under the Apache License, Version 2.0 (the "License").
+* You may not use this file except in compliance with the License.
+* A copy of the License is located at
+*
+*  http://aws.amazon.com/apache2.0
+*
+* or in the "license" file accompanying this file. This file is distributed
+* on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+* express or implied. See the License for the specific language governing
+* permissions and limitations under the License.
+*/
+
+#include <assert.h>
+
+#include <aws/common/common.h>
+
+struct aws_memory_pool {
+    struct aws_allocator *alloc;
+    size_t arena_size;
+    size_t element_size;
+    uint8_t *arena;
+    void *free_list;
+    int overflow_count;
+};
+
+struct aws_memory_pool *aws_memory_pool_init(struct aws_allocator* alloc, size_t element_size, int element_count) {
+    size_t stride = element_size > sizeof(void *) ? element_size : sizeof(void *);
+    size_t arena_size = sizeof(struct aws_memory_pool) + stride * element_count;
+    struct aws_memory_pool *pool = (struct aws_memory_pool *)aws_mem_acquire(alloc, arena_size);
+
+    if (!pool) {
+        return NULL;
+    }
+
+    pool->alloc = alloc;
+    pool->arena_size = arena_size - sizeof(struct aws_memory_pool);
+    pool->element_size = element_size;
+    pool->arena = (uint8_t *)(pool + 1);
```

Comment:
Didn't write this with alignment in mind -- can add this feature in at another time.

Reply:
@bdonlan I thought about alignment too when reviewing this code, but found all the types already aligned on 8/4 byte boundary. Also, the type used in this memory pool is composed of plain pointers (no doubles, or long double) so I held my mouth shut.

Can you elaborate on the SIMD alignment requirements with an example or something? I'd love to learn more.

Reply:
oh yeah good point. I see xmm registers used quite frequently even without touching floating-points. For some dumb reason I always tied SIMD to usage of compiler intrinsics such as __mm_* functions.

Reply:
When the compiler implicitly uses SIMD for FP or integer operations it'll use the unaligned variants unless it can prove 128-bit (or higher) alignment. However there are direct SIMD intrinsics that need alignments up to 512 bits.